packages/app/app/actions/queue.ts

```diff
@@ -262,11 +264,23 @@ export function addPlaylistTracksToQueue(tracks) {
 function dispatchWithShuffle(dispatch, getState, action) {
   const state = getState();
   const settings = state.settings;
-  const queue = state.queue;
+  let queue = state.queue;
 
   if (settings.shuffleQueue) {
-    const index = _.random(0, queue.queueItems.length - 1);
-    dispatch(selectSong(index));
+    let unplayedQueue = queue.queueItems.filter(item => !item.played);
+    if (unplayedQueue.length === 0){
+      queue.queueItems.map(item => dispatch(updateQueueItem({ ... item, played: false})));
+    }
+    if (queue.currentSong !== null) {
+      const currentSongUuid = queue.queueItems[queue.currentSong].uuid;
+      dispatch(updateQueueItem({ ...queue.queueItems[queue.currentSong], played: true, uuid: currentSongUuid }));
+      queue = getState().queue;
+    }
+    unplayedQueue = queue.queueItems.filter(item => !item.played);
```

Comment:
What happens when all the tracks have this flag set to true?

Reply:
if that happens line 272 will run which is `if (unplayedQueue.length === 0) { ... }`